### PR TITLE
Add debounce_job primitive (v0.2.0)

### PR DESCRIPTION
**Pull request by szhuravski-fp**
## What

Adds `debounce_job` as a third enqueue-control primitive alongside `limit_queue` and `throttle_job`.

Trailing-edge debounce semantics: for any burst of `perform_later` triggers on a given resource, exactly one execution fires after `duration` has elapsed since the **last** trigger. Coalesces the burst rather than firing on the leading edge.

**How it works:**
- Every `perform_later` is a trigger, not a direct enqueue (`job_id` is set to `nil`).
- The first trigger in a burst atomically claims a scheduled slot in Redis and enqueues one internal delayed job (`duration` from now).
- Subsequent triggers overwrite the target epoch in Redis (`quiet_until`) and are dropped — no additional Sidekiq job.
- When the delayed job wakes up, check the `quiet_until`: if it has passed, execute; if a newer trigger extended it, reschedule and drop.

Atomicity is guaranteed by two Lua scripts in `SidekiqAdapter` to prevent races between concurrent triggers.

**Version bump:** `0.1.1` → `0.2.0`
**Testing setup:** Added testing Sidekiq adapter against actual Redis

## Why

Throttle fires on the leading edge; debounce fires on the trailing edge. This is the right primitive for workflows where you want to act once after a burst settles — search re-indexing after rapid model updates, webhook flood collapsing, etc.

### CODE REVIEW

**Review by june-sky — COMMENTED** [review]
@claude review this PR

**Comment by june-sky** [comment]
Moving the `Sidekiq/Redis` configuration into `spec_helper.rb` means every spec file now requires a live Redis connection at load time — including limiter_spec.rb and limiter_throttle_spec.rb, which have no Redis dependency at all. Any contributor running bundle exec rspec without Docker will get a Redis::CannotConnectError on tests that were previously self-contained.
Please move the require `sidekiq` and `Sidekiq.configure_*` calls into `sidekiq_adapter_debounce_spec.rb` directly. That way the integration tests declare their own infrastructure dependency, and unit tests stay runnable without Redis.

```
bundle exec rspec spec/

Finished in 0.05566 seconds (files took 0.23544 seconds to load)
31 examples, 11 failures

Failed examples:

rspec ./spec/active_job/sidekiq_adapter_debounce_spec.rb:42 # ActiveJob::Limiter::QueueAdapters::SidekiqAdapter debounce #register_debounce_trigger returns true on the first trigger and creates both keys
rspec ./spec/active_job/sidekiq_adapter_debounce_spec.rb:54 # ActiveJob::Limiter::QueueAdapters::SidekiqAdapter debounce #register_debounce_trigger returns false on subsequent triggers (slot already taken)
rspec ./spec/active_job/sidekiq_adapter_debounce_spec.rb:59 # ActiveJob::Limiter::QueueAdapters::SidekiqAdapter debounce #register_debounce_trigger slides quiet_until forward on each trigger
rspec ./spec/active_job/sidekiq_adapter_debounce_spec.rb:70 # ActiveJob::Limiter::QueueAdapters::SidekiqAdapter debounce #register_debounce_trigger sets TTL on both keys equal to duration
rspec ./spec/active_job/sidekiq_adapter_debounce_spec.rb:82 # ActiveJob::Limiter::QueueAdapters::SidekiqAdapter debounce #register_debounce_trigger does not refresh the scheduled key TTL on subsequent triggers
rspec ./spec/active_job/sidekiq_adapter_debounce_spec.rb:96 # ActiveJob::Limiter::QueueAdapters::SidekiqAdapter debounce #claim_debounce_execution when quiet_until key is absent returns a claimed DebounceClaim and leaves no keys
rspec ./spec/active_job/sidekiq_adapter_debounce_spec.rb:111 # ActiveJob::Limiter::QueueAdapters::SidekiqAdapter debounce #claim_debounce_execution when quiet_until is in the past returns a claimed DebounceClaim
rspec ./spec/active_job/sidekiq_adapter_debounce_spec.rb:117 # ActiveJob::Limiter::QueueAdapters::SidekiqAdapter debounce #claim_debounce_execution when quiet_until is in the past deletes both keys
rspec ./spec/active_job/sidekiq_adapter_debounce_spec.rb:136 # ActiveJob::Limiter::QueueAdapters::SidekiqAdapter debounce #claim_debounce_execution when quiet_until is in the future returns an unclaimed DebounceClaim with wait_seconds approximating remaining time
rspec ./spec/active_job/sidekiq_adapter_debounce_spec.rb:142 # ActiveJob::Limiter::QueueAdapters::SidekiqAdapter debounce #claim_debounce_execution when quiet_until is in the future extends the is_job_scheduled TTL to approximately remaining time + DEBOUNCE_TTL_BUFFER
rspec ./spec/active_job/sidekiq_adapter_debounce_spec.rb:150 # ActiveJob::Limiter::QueueAdapters::SidekiqAdapter debounce #claim_debounce_execution when quiet_until is in the future keeps both keys alive
```

**Review by june-sky — APPROVED** [review]
The implementation is well-structured and the Lua scripts look correct, but I'd recommend validating `debounce_job` on one of your own jobs in production for a week or two before merging it into the shared library — this gem is used across production jobs, and edge cases like zombie lock (up to duration + 1h recovery window if a waker job is lost) and potential double execution are hard to catch in tests but can surface under real load. Once battle-tested, adding it to the shared lib is a much safer proposition.

File: lib/active_job/limiter/queue_adapters/sidekiq_adapter.rb [hunk 1]
```diff
@@ -40,6 +43,77 @@ def release_lock_for_job_resource(name, job, resource_id)
...
+          DEBOUNCE_TRIGGER_SCRIPT = <<~LUA.freeze
+            local quiet_until_key      = KEYS[1]  -- stores the earliest epoch at which execution is allowed
+            local is_job_scheduled_key = KEYS[2]  -- 1 iff a delayed job is already waiting in Sidekiq
+            local new_quiet_until      = ARGV[1]  -- Time.now + duration (float seconds, as string)
+            local ttl                  = ARGV[2]  -- integer seconds until both keys expire
+
+            -- Always push quiet_until forward; EX = set expiration in seconds
+            redis.call('SET', quiet_until_key, new_quiet_until, 'EX', ttl)
+            -- NX = only set if key does not exist; returns 'OK' on success, means this caller won the scheduling slot
+            return redis.call('SET', is_job_scheduled_key, '1', 'NX', 'EX', ttl)
```
Comment (MarkKiepeFlexport) [inline lib/active_job/limiter/queue_adapters/sidekiq_adapter.rb:75]: Never did I expect to see Lua at Flexport, this is awesome

**Review by MarkKiepeFlexport — APPROVED** [review]

**Comment by szhuravski-fp** [comment]
I have tested it for more than a week with `ShipmentCustomsEntryPacketGeneratorJob` job. Looks good
